### ISSUE TITLE
No Simulation Message

### ISSUE THREAD

**Issue by boverhof** [CCSI-Toolset/FOQUS#1114]
# Description
Need a better no simulation message.
- Current Message:  'The simulation completed with an error 20'
- Issue:  Didn't load the simulation into Turbine
- Need to look at logs in order to understand what happened.

- [ ] Change Message in pop up "Simulation {{ name }} is Missing, goto 'Add/upload model to Turbine under session "

## Steps to Reproduce
- Load up a flowsheet that contains a node with a model [BFB](https://github.com/CCSI-Toolset/FOQUS/blob/master/examples/test_files/OUU_Remote_Turbine_BFB/BFB_OUU_06_09_2020.foqus)
- Hit "Run"

```
Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\envs\foqus\lib\site-packages\foqus_lib\framework\sim\turbineConfiguration.py", line 313, in getSimApplication
    raise TurbineInterfaceEx(
foqus_lib.framework.sim.turbineConfiguration.TurbineInterfaceEx: 360 - Could not find simulation, Could not find simulation: HighConcentration





During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\envs\foqus\lib\threading.py", line 932, in _bootstrap_inner
    self.run()
  File "C:\ProgramData\Anaconda3\envs\foqus\lib\site-packages\foqus_lib\framework\graph\graph.py", line 882, in run
    _log.exception("Exception in graph thread: %s" % (self.ex))
TypeError: not all arguments converted during string formatting
2023-03-06 20:14:09,118 - ERROR - foqus.foqus_lib.gui.main.mainWindow - to results graph thread was likely terminated
```

<img width="975" alt="Screen Shot 2023-03-06 at 11 14 24 AM" src="https://user-images.githubusercontent.com/738295/223208358-22bb2ab9-118b-424b-b5ef-7f5075eb148c.png">

**Comment by kbuma** [comment]
I did a clean install on MacOS -> ran foqus -> loaded the example you listed -> tried to run it

It looks like it tried to connect to a local turbine instance (which I don't have running as I am on Mac OS) and provided a different cryptic error.

![image](https://github.com/CCSI-Toolset/FOQUS/assets/2824957/43419c04-a0fb-49aa-a913-866d38e67621)

Traceback (most recent call last):
  File "/Users/kberket/anaconda3/envs/ccsi-foqus/lib/python3.8/site-packages/foqus_lib/framework/graph/node.py", line 1392, in runPython
    exec(self.pythonCode)
  File "<string>", line 4, in <module>
  File "/Users/kberket/anaconda3/envs/ccsi-foqus/lib/python3.8/site-packages/foqus_lib/framework/graph/node.py", line 1325, in runModel
    self.runTurbineCalc(retry=self.options["Retry"].value)
  File "/Users/kberket/anaconda3/envs/ccsi-foqus/lib/python3.8/site-packages/foqus_lib/framework/graph/node.py", line 1515, in runTurbineCalc
    configProfile.startConsumer(self.name, self.modelName)
  File "/Users/kberket/anaconda3/envs/ccsi-foqus/lib/python3.8/site-packages/foqus_lib/framework/sim/turbineConfiguration.py", line 340, in startConsumer
    app = self.getSimApplication(simName)
  File "/Users/kberket/anaconda3/envs/ccsi-foqus/lib/python3.8/site-packages/foqus_lib/framework/sim/turbineConfiguration.py", line 322, in getSimApplication
    raise TurbineInterfaceEx(
  File "/Users/kberket/anaconda3/envs/ccsi-foqus/lib/python3.8/site-packages/foqus_lib/framework/sim/turbineConfiguration.py", line 106, in __init__
    elif e.reason.errno == 10060:
AttributeError: 'ConnectionError' object has no attribute 'reason'
2023-06-12 13:24:55,118 - ERROR - foqus.foqus_lib.framework.graph.graph - runGraph(BFB): calcError=21
2023-06-12 13:24:55,118 - DEBUG - foqus.foqus_lib.framework.graph.graph - solve: no tears flowsheet done run post solve nodes
2023-06-12 13:24:55,585 - DEBUG - foqus.foqus_lib.gui.main.mainWindow - stopSim SingleRun: {'solTime': 0.009353876113891602, 'input': {'graph': OrderedDict(), 'BFB': OrderedDict([('BFBadsB.Dt', 11.897), ('BFBadsB.dx', 0.0127), ('BFBadsB.Lb', 2.085), ('BFBadsM.Dt', 15.0), ('BFBadsM.dx', 0.06695), ('BFBadsM.Lb', 1.972), ('BFBadsT.Dt', 15.0), ('BFBadsT.dx', 0.062397), ('BFBadsT.Lb', 2.203), ('BFBRGN.Dt', 9.041), ('BFBRGN.Lb', 8.886), ('BFBRGNTop.Dt', 9.195), ('BFBRGNTop.Lb', 7.1926), ('GHXfg.A_exch', 16358.0), ('Kd', 100.0), ('BFBadsB.Cr', 1.0), ('BFBadsM.Cr', 1.0), ('BFBadsT.Cr', 1.0), ('BFBRGN.Cr', 1.0), ('BFBRGNTop.Cr', 1.0), ('dp', 0.00015), ('GHXfg.GasIn.P', 1.01325), ('GHXfg.GasIn.T', 54.0), ('fg_flow', 100377.0)])}, 'output': {'graph': OrderedDict([('error', 1.0)]), 'BFB': OrderedDict([('Cost_coe', 138.8242573988463), ('Cost_coe_obj', 138.8242573988463), ('Cost_ads', 34832226.20639935), ('Cost_aux_power', 23385.503427224274), ('Cost_op_cooling_water', 12562503.664959071), ('Cost_op_cooling_water_flow', 2923843.78423705), ('Cost_op_fixed', 54527170.10507197), ('Cost_op_var', 172522109.68218017), ('Cost_rgn', 16323489.589995965), ('Cost_shx', 39020395.75661277), ('Cost_steam_power', 178992.84008439552), ('Cost_steam_tot', 695217.1196293444), ('Cost_toc', 2111605826.971341), ('Cost_toc_sorb', 84011457.36869758), ('removalCO2', 0.8996801078092209), ('SHX.CWFR', 26470.78926712216), ('SHX.LeanIn.T', 146.3976529163801), ('SHX.RichIn.T', 84.06464787557616), ('SHX.SteamFR', 37652.63814666193), ('removalCO2_slack', 0.0), ('slugab_slack', 0.0), ('slugam_slack', 0.0), ('slugat_slack', 0.0), ('slugrb_slack', 0.0), ('slugrt_slack', 0.0), ('BFBRGN.GasIn.F', 1357.089941010187), ('F_solids', 12702449.02608155), ('GHXfg.HXIn.F', 135603.6604814625), ('SHX.LeanOut.T', 71.01606811426312), ('SHX.RichOut.T', 170.0), ('status', 0.0)])}, 'graphError': 1, 'nodeError': {'BFB': 21}, 'nodeSettings': {'BFB': {'Visible': False, 'Initialize Model': False, 'Reset': False, 'Reset on Fail': True, 'Retry': False, 'Allow Simulation Warnings': True, 'Max consumer reuse': 90, 'Maximum Wait Time (s)': 1440.0, 'Maximum Run Time (s)': 840.0, 'Min Status Check Interval': 4.0, 'Max Status Check Interval': 5.0, 'Override Turbine Configuration': '', 'RunMode': 'Steady State', 'printlevel': 0, 'homotopy': 0, 'Snapshot': '', 'TimeSeries': [0.0], 'Script': 'Run_Steady', 'MinStepSize': 0.001, 'TimeUnits': 'Hours'}}, 'turbineMessages': {'BFB': ''}, 'input_vectorvals': {'BFB': OrderedDict()}, 'output_vectorvals': {'BFB': OrderedDict()}}

**Comment by kbuma** [comment]
@boverhof what OS and other CCSI systems were you running when you got the error? i would like to be able to reproduce that same error.

**Comment by boverhof** [comment]
I was running it on a Windows 2016 server.  But that error is in scope for this issue.  It's slightly more informative but "error 1" doesn't mean much.

**Comment by kbuma** [comment]
For the "error 1" that I was getting on the Mac this looks to be appropriate since once I clicked on the node I got a better message in the resulting dialog.

![image](https://github.com/CCSI-Toolset/FOQUS/assets/2824957/ea06a93f-661b-4814-ad45-81b5a8d46cbe)

Also, the logs are pretty clear about the fact that it couldn't connect to the Turbine server.

**Comment by kbuma** [comment]
From reviewing the code, error 20 indicates that somehow the graph thread was terminated. It looks like in that case it would be difficult to provide more detail in the dialog.  Maybe we can add messaging to check the logs in order to try to get more information.

**Comment by boverhof** [comment]
I was thinking we might just add something to this function to get a more meaningful message:

```python
class TurbineInterfaceEx(foqusException):
    """
    This is an exception class for FOQUS interaction with Turbine, it also tries
    to sort out exactly what the problem is if there is any trouble connecting to
    Turbine.
    """

    ...
    def setCodeStrings(self):
        self.codeString[0] = "Unknown Error"
        self.codeString[1] = "Unknown gateway URL error"
        self.codeString[2] = (
            "Gateway did not respond (maybe a bad address, "
            "network problem, or gateway is down)"
        )
        self.codeString[3] = "Forcible connection reset by Gateway"
        self.codeString[5] = "Gateway SSL error (probably wrong address)"
        self.codeString[7] = "Unknown Socket Error"
        self.codeString[10] = "Gateway unknown http error"
        self.codeString[
            11
        ] = "Gateway 401 authentication

```

**Comment by boverhof** [comment]
Okay that sounds reasonable.   I think I see what you're saying.  I think the error "20" here is the index, and the message is what I am seeing but what else could we know?  

`foqus-boverhof boverhof$ sed -n '47,65p'  foqus_lib//framework/graph/graph.py`
```python
class GraphEx(foqusException):
    def setCodeStrings(self):
        self.codeString[0] = "Finished Normally"
        self.codeString[1] = "Failure in node calculation, " "check node status"
        self.codeString[3] = (
            "Failure to create worker node " "(most likely could not talk to Turbine)"
        )
        self.codeString[5] = "Specified unknown tear solver"
        self.codeString[11] = "Wegstein solver failed to converge"
        self.codeString[12] = "Direct solver failed to converge"
        self.codeString[16] = "Error in presolve node"
        self.codeString[17] = "Error in postsolve node"
        self.codeString[19] = "Exception during graph execution (see foqus.log)"
        self.codeString[20] = "Flowsheet thread terminated"
```

**Comment by kbuma** [comment]
Monitoring the running of a simulation from the main window is currently implemented using polling, ie a `QTimer` that fires off `checkSim` on a regular basis. The check looks at the current state of the run as it's been updated by the underlying logic and if there's an error, stops the simulation and presents the error dialog. (This error display check is also performed when the stop button is pressed) The error message that is generated is completely based on the integer value of `errorStat` in the `Graph` object (`dat.flowsheet` in `mainWindow`). 

Because of this, what is presented to the user is the most recent error (which may be a side-effect of the original cause).

In order to avoid that we would need to refactor the UX error handling code to an event based system (e.g. using Qt signals and slots). This would be a significant effort as it would likely require a refactor of the entire UX.

Thus, I propose the following mitigations to enhance the usability of the error UX:

1. add documentation on how users can view (and/or share) the application logs. The application logs provide information about all the exceptions and errors encountered during the application runtime and can help pinpoint the cause of the error.
2. improve some of the error dialog messaging, including providing a link to the documentation about accessing logs

**Comment by boverhof** [comment]
Think we should clean up the "other" error we are hitting before moving on. 

- [ ]  add check if hasattr(e, 'reason')  and do the appropriate action:

```
File "/Users/kberket/anaconda3/envs/ccsi-foqus/lib/python3.8/site-packages/foqus_lib/framework/sim/turbineConfiguration.py", line 106, in init
elif e.reason.errno == 10060:
AttributeError: 'ConnectionError' object has no attribute 'reason'
```

**Comment by lbianchi-lbl** [comment]
This is addressed as a "low-hanging fruit" by #1149. A more complete way to handle errors would require considerably more effort, possibly involving switching to using Qt's built-in signal and slots functionality.
